File: flutter/lib/utils/relay.dart
// Message relay: the app-layer hop logic that turns Nearby's one-room P2P
// clustering into something that reaches past a single radio range.
//
// Pure functions only — no native calls, no store access — so the hop rules can
// be reasoned about and tested on their own. This is the part that decides
// whether A -> B -> C works, so it stays boring and inspectable.
//
// Port of src/utils/relay.ts.
import 'dart:convert';

/// What actually travels between phones. Kept small; Nearby sends it as text.
enum EnvelopeKind {
  msg,
  coin;

  String get wire => name;

  static EnvelopeKind fromWire(String value) =>
      value == 'coin' ? EnvelopeKind.coin : EnvelopeKind.msg;
}

const int defaultTtl = 3;

class Envelope {
  /// Unique per original message. Dedupe key — survives every hop unchanged.
  final String id;

  /// Device id of whoever composed it. Never rewritten by a relay.
  final String from;

  /// Device id of the intended reader, or a thread id for a group.
  final String to;

  final EnvelopeKind kind;

  /// Opaque to relays. Ciphertext once the vault lands.
  final String body;

  /// Hops remaining. A relay forwards only while this is above zero.
  final int ttl;

  /// Device ids this has already passed through, in order. Builds the route strip.
  final List<String> path;

  /// Composed-at, from the sender's clock. Display only — never trusted for ordering.
  final int at;

  const Envelope({
    required this.id,
    required this.from,
    required this.to,
    required this.kind,
    required this.body,
    required this.at,
    this.ttl = defaultTtl,
    this.path = const [],
  });

  Envelope copyWith({int? ttl, List<String>? path}) => Envelope(
    id: id,
    from: from,
    to: to,
    kind: kind,
    body: body,
    at: at,
    ttl: ttl ?? this.ttl,
    path: path ?? this.path,
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'from': from,
    'to': to,
    'kind': kind.wire,
    'body': body,
    'ttl': ttl,
    'path': path,
    'at': at,
  };
}

Envelope newEnvelope({
  required String id,
  required String from,
  required String to,
  required EnvelopeKind kind,
  required String body,
  required int at,
  int? ttl,
}) => Envelope(
  id: id,
  from: from,
  to: to,
  kind: kind,
  body: body,
  at: at,
  ttl: ttl ?? defaultTtl,
);

/// Hops taken so far. 0 means it arrived directly from the sender.
int hopsTaken(Envelope e) => e.path.length;

/// The device that handed it to us, if anyone relayed it.
String? relayedBy(Envelope e) => e.path.isEmpty ? null : e.path.last;

/// Fixed-size ring of message ids we've already handled. A plain Set would grow
/// without bound over a long demo; this forgets the oldest instead.
class SeenCache {
  SeenCache([this.limit = 512]);

  final int limit;
  final List<String> _order = [];
  final Set<String> _ids = {};

  /// True if this id was already seen. Records it either way.
  bool check(String id) {
    if (_ids.contains(id)) return true;
    _ids.add(id);
    _order.add(id);
    if (_order.length > limit) {
      final oldest = _order.removeAt(0);
      _ids.remove(oldest);
    }
    return false;
  }

  int get size => _ids.length;
}

sealed class Decision {
  const Decision();
}

class DropDecision extends Decision {
  const DropDecision(this.why);

  /// 'duplicate' | 'expired' | 'loop'
  final String why;
}

class DeliverDecision extends Decision {
  const DeliverDecision(this.envelope);

  final Envelope envelope;
}

class RelayDecision extends Decision {
  const RelayDecision(this.envelope, {this.excludePeer});

  final Envelope envelope;
  final String? excludePeer;
}

/// The whole hop rule, in one place.
///
/// `me` is this device's id, `fromPeer` the peer that handed it over (null if we
/// composed it). Returns what to do — the caller performs the side effects.
Decision route(Envelope envelope, String me, SeenCache seen, {String? fromPeer}) {
  if (seen.check(envelope.id)) return const DropDecision('duplicate');

  // Our own id already in the path means it came back around to us.
  if (envelope.path.contains(me)) return const DropDecision('loop');

  if (envelope.to == me) return DeliverDecision(envelope);

  if (envelope.ttl <= 0) return const DropDecision('expired');

  // Carry it on: burn a hop and record that we touched it, so the recipient can
  // see the route it took.
  return RelayDecision(
    envelope.copyWith(ttl: envelope.ttl - 1, path: [...envelope.path, me]),
    excludePeer: fromPeer,
  );
}

/// Group messages are delivered to everyone AND relayed onward.
bool isGroup(String to) => to.startsWith('g:');

String encode(Envelope e) => jsonEncode(e.toJson());

/// Never throws — a malformed payload from another build must not crash the app.
Envelope? decode(String text) {
  try {
    final e = jsonDecode(text);
    if (e is! Map ||
        e['id'] is! String ||
        e['from'] is! String ||
        e['to'] is! String ||
        e['body'] is! String ||
        e['ttl'] is! num ||
        e['path'] is! List ||
        (e['path'] as List).any((p) => p is! String)) {
      return null;
    }
    return Envelope(
      id: e['id'] as String,
      from: e['from'] as String,
      to: e['to'] as String,
      kind: EnvelopeKind.fromWire(e['kind'] as String? ?? 'msg'),
      body: e['body'] as String,
      ttl: (e['ttl'] as num).toInt(),
      path: List<String>.from(e['path'] as List),
      at: (e['at'] as num?)?.toInt() ?? 0,
    );
  } catch (_) {
    return null;
  }
}

File: src/utils/relay.ts
// Message relay: the app-layer hop logic that turns Nearby's one-room P2P
// clustering into something that reaches past a single radio range.
//
// Pure functions only — no native calls, no store access — so the hop rules can
// be reasoned about and tested on their own. This is the part that decides
// whether A -> B -> C works, so it stays boring and inspectable.

/** What actually travels between phones. Kept small; Nearby sends it as text. */
export type Envelope = {
  /** Unique per original message. Dedupe key — survives every hop unchanged. */
  id: string;
  /** Device id of whoever composed it. Never rewritten by a relay. */
  from: string;
  /** Device id of the intended reader, or a thread id for a group. */
  to: string;
  /**
   * `revert` carries the id of an earlier coin message to undo.
   * `image` carries base64 image data, split across parts.
   * `event` carries a calendar event as JSON.
   */
  kind: 'msg' | 'coin' | 'revert' | 'image' | 'event';
  /** Opaque to relays. Ciphertext once the vault lands. */
  body: string;
  /** Set on every part of a split payload. Shared by all parts of one message. */
  gid?: string;
  /** Which part this is, and how many there are. Absent means a whole message. */
  part?: { i: number; n: number };
  /** Hops remaining. A relay forwards only while this is above zero. */
  ttl: number;
  /** Device ids this has already passed through, in order. Builds the route strip. */
  path: string[];
  /** Composed-at, from the sender's clock. Display only — never trusted for ordering. */
  at: number;
};

export const DEFAULT_TTL = 3;

export const newEnvelope = (
  args: Pick<Envelope, 'id' | 'from' | 'to' | 'kind' | 'body' | 'at'> & { ttl?: number }
): Envelope => ({
  ...args,
  ttl: args.ttl ?? DEFAULT_TTL,
  path: [],
});

/** Hops taken so far. 0 means it arrived directly from the sender. */
export const hopsTaken = (e: Envelope): number => e.path.length;

/** The device that handed it to us, if anyone relayed it. */
export const relayedBy = (e: Envelope): string | undefined => e.path[e.path.length - 1];

/**
 * Fixed-size ring of message ids we've already handled. A plain Set would grow
 * without bound over a long demo; this forgets the oldest instead.
 */
export class SeenCache {
  private order: string[] = [];
  private set = new Set<string>();
  constructor(private limit = 512) {}

  /** True if this id was already seen. Records it either way. */
  check(id: string): boolean {
    if (this.set.has(id)) return true;
    this.set.add(id);
    this.order.push(id);
    if (this.order.length > this.limit) {
      const oldest = this.order.shift()!;
      this.set.delete(oldest);
    }
    return false;
  }

  get size() {
    return this.set.size;
  }
}

export type Decision =
  | { action: 'drop'; why: 'duplicate' | 'expired' | 'loop' }
  | { action: 'deliver'; envelope: Envelope }
  | { action: 'relay'; envelope: Envelope; excludePeer?: string };

/**
 * The whole hop rule, in one place.
 *
 * `me` is this device's id, `from` the peer that handed it over (undefined if we
 * composed it). Returns what to do — the caller performs the side effects.
 */
export function route(
  envelope: Envelope,
  me: string,
  seen: SeenCache,
  fromPeer?: string
): Decision {
  if (seen.check(envelope.id)) return { action: 'drop', why: 'duplicate' };

  // Our own id already in the path means it came back around to us.
  if (envelope.path.includes(me)) return { action: 'drop', why: 'loop' };

  if (envelope.to === me) return { action: 'deliver', envelope };

  if (envelope.ttl <= 0) return { action: 'drop', why: 'expired' };

  // Carry it on: burn a hop and record that we touched it, so the recipient can
  // see the route it took.
  return {
    action: 'relay',
    envelope: { ...envelope, ttl: envelope.ttl - 1, path: [...envelope.path, me] },
    excludePeer: fromPeer,
  };
}

/** Group messages are delivered to everyone AND relayed onward. */
export const isGroup = (to: string) => to.startsWith('g:');

/**
 * Nearby sends a BYTES payload, which Google caps at 32 KiB. Anything larger —
 * an image, realistically — has to travel as parts and be put back together at
 * the far end. Relays forward parts individually and never need to understand
 * them.
 */
export const MAX_PAYLOAD = 32 * 1024;

/** Leaves room for the JSON envelope wrapped around each part. */
export const CHUNK_CHARS = 24_000;

export function splitBody(body: string, size = CHUNK_CHARS): string[] {
  if (body.length <= size) return [body];
  const parts: string[] = [];
  for (let i = 0; i < body.length; i += size) parts.push(body.slice(i, i + size));
  return parts;
}

/**
 * Collects parts until a message is whole. Incomplete groups are dropped once
 * the cache is full rather than held forever — a sender that walks away
 * mid-image must not leak memory.
 */
export class Reassembler {
  private groups = new Map<string, { parts: (string | undefined)[]; n: number }>();
  constructor(private limit = 16) {}

  /** Returns the full body once every part has arrived, otherwise null. */
  add(e: Envelope): string | null {
    if (!e.part || !e.gid) return e.body;

    let group = this.groups.get(e.gid);
    if (!group) {
      // Filled, not sparse: `.some` skips holes in a sparse array, which would
      // report a transfer complete on its first part.
      group = { parts: new Array(e.part.n).fill(undefined), n: e.part.n };
      this.groups.set(e.gid, group);
      // Evict the oldest partial group if we're tracking too many.
      if (this.groups.size > this.limit) {
        const oldest = this.groups.keys().next().value;
        if (oldest !== undefined) this.groups.delete(oldest);
      }
    }

    group.parts[e.part.i] = e.body;
    if (group.parts.some((p) => p === undefined)) return null;

    this.groups.delete(e.gid);
    return group.parts.join('');
  }

  get pending() {
    return this.groups.size;
  }
}

export const encode = (e: Envelope): string => JSON.stringify(e);

/** Never throws — a malformed payload from another build must not crash the app. */
export function decode(text: string): Envelope | null {
  try {
    const e = JSON.parse(text);
    if (
      typeof e?.id !== 'string' ||
      typeof e?.from !== 'string' ||
      typeof e?.to !== 'string' ||
      typeof e?.body !== 'string' ||
      typeof e?.ttl !== 'number' ||
      !Array.isArray(e?.path)
    ) {
      return null;
    }
    return e as Envelope;
  } catch {
    return null;
  }
}
